packages/loopring_v3/contracts/core/impl/ExchangeV3.sol

```diff
@@ -554,17 +554,34 @@ contract ExchangeV3 is IExchangeV3
         address to,
         address token,
         uint96  amount,
-        uint32  nonce,
+        uint32  storageID,
         address newRecipient
         )
         external
         override
         nonReentrant
         onlyFromUserOrAgent(from)
     {
+        require(newRecipient != address(0), "INVALID_DATA");
```

Comment:
should we allow the recipient to be set to 0? I'm thinking of the following scenario:
- The withdrawal is done, the recipient is set to `withdrawal.to`
- The FastWithdrawalContract reads the recipient, find it's non-zero, then set the new value to 0, to get some gas refund, then abord the operation (but still mark the approval as used)

Reply:
> (but still mark the approval as used)

I guess this is the problem. Unless time (or something similar) is used somehow, somewhere we need to store that the fast withdrawal approval cannot be used any more. With the current method we only need to depend on the stored recipient for that (if non-zero it cannot be used), no other hash/nonce needs to be stored elsewhere. Otherwise we reset storage to zero in one place, but need to store a non-zero value elsewhere, which makes things more expensive.

Reply:
Also the recipient can be set from the account owner himself, or any of it's agents. So it's important that it cannot be overridden because the owner can directly do that himself or just register another agent contract that can do it as well. Something to keep in mind as well. Which is why in the opening post I stored the "store until" timestamp in the protocol itself, because we cannot depend on the logic inside a single agent to keep it safe.

We could make it so that the data with the address that set it, is also the only that address can reset it again, but that makes things a bit more complex again (and will probably also make it so we need to use 2 storage slots inside of the 1 we need now to store that extra address).

Reply:
I was thinking the FastWithdrawal contract will also do a bookkeeping. If our solution can make it unnecessary, that's even better.

Reply:
Currently none is needed, no storage is currently used for fast withdrawals in the FastWithdrawalAgent contract.